starboard/shared/libiamf/iamf_config_reader.cc

```diff
@@ -0,0 +1,531 @@
+  *ptr = ByteSwap(*ptr);
+  pos_ += sizeof(uint32_t);
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::ReadLeb128(uint32_t* ptr) {
+  if (!HasBytes(sizeof(uint32_t)) || !ptr) {
+    error_ = true;
+    return false;
+  }
+  int bytes_read = ReadLeb128Internal(buf_ + pos_, ptr);
+  if (bytes_read < 0) {
+    error_ = true;
+    return false;
+  }
+  pos_ += bytes_read;
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::ReadString(std::string* str) {
+  int bytes_read = ReadStringInternal(buf_ + pos_, str);
+  if (bytes_read < 0) {
+    error_ = true;
+    return false;
+  }
+  pos_ += bytes_read;
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::SkipBytes(size_t size) {
+  if (!HasBytes(size)) {
+    error_ = true;
+    return false;
+  }
+  pos_ += size;
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::SkipLeb128() {
+  uint32_t val;
+  return ReadLeb128(&val);
+}
+
+bool IamfConfigReader::BufferReader::SkipString() {
+  std::string str;
+  return ReadString(&str);
+}
+
+int IamfConfigReader::BufferReader::ReadLeb128Internal(const uint8_t* buf,
+                                                       uint32_t* value) {
+  SB_DCHECK(buf);
+  SB_DCHECK(value);
+  *value = 0;
+  bool error = true;
+  size_t i = 0;
+  for (; i < sizeof(uint32_t); ++i) {
+    uint8_t byte = buf[i];
+    *value |= ((byte & 0x7f) << (i * 7));
+    if (!(byte & 0x80)) {
+      error = false;
+      break;
+    }
+  }
+
+  if (error) {
+    return -1;
+  }
+  return i + 1;
+}
+
+int IamfConfigReader::BufferReader::ReadStringInternal(const uint8_t* buf,
+                                                       std::string* str) {
+  SB_DCHECK(buf);
+
+  int remaining_size = static_cast<int>(size_) - pos_;
+  if (remaining_size <= 0) {
+    return -1;
+  }
+
+  // The size of the string is capped to 128 bytes.
+  int max_str_size = std::min(remaining_size, 128);
+  str->clear();
+
+  size_t size = 0;
+  while (buf[size] != '\0' && size < max_str_size) {
+    size++;
+  }
+
+  if (buf[size] != '\0') {
+    return -1;
+  }
+
+  if (size > 0) {
+    str->resize(size);
+    std::memcpy(str->data(), reinterpret_cast<const char*>(buf), size);
+  }
+
+  // Account for null terminator byte.
+  return ++size;
+}
+
+IamfConfigReader::IamfConfigReader(
+    const scoped_refptr<InputBuffer>& input_buffer,
+    bool prefer_binaural_audio,
+    bool prefer_surround_audio)
+    : prefer_binaural_audio_(prefer_binaural_audio),
+      prefer_surround_audio_(prefer_surround_audio) {
+#if SB_IS_BIG_ENDIAN
+#error IamfConfigReader assumes little-endianness.
+#endif  // SB_IS_BIG_ENDIAN
+  SB_DCHECK(!(prefer_binaural_audio && prefer_surround_audio));
+  Read(input_buffer);
+}
+
+bool IamfConfigReader::Read(const scoped_refptr<InputBuffer>& input_buffer) {
+  SB_DCHECK(input_buffer->data());
+  BufferReader reader(input_buffer->data(), input_buffer->size());
+
+  while (!is_valid() && reader.pos() < reader.size()) {
+    RCHECK(ReadOBU(&reader));
```

Comment:
The code is to read OBU until `mix_presentation_id` is found. Let's make the code more clear on that.